### PR TITLE
Revamped configs for dynamic parameters

### PR DESCRIPTION
**Pull request by thundron**
Change-type: major
Signed-off-by: Lorenzo Alberto Maria Ambrosi <lorenzothunder.ambrosi@gmail.com>

### CODE REVIEW

**Comment by thundron** [comment]
@cskiraly @nopbyte Please review what I wrote and while doing so, make any suggestions you might find useful so we can discuss them.

**Comment by nopbyte** [comment]
So, the idea is to not use Swagger after all for the API specification?  :(

**Comment by thundron** [comment]
@nopbyte Given how the API specs themselves are actually written, it would force us to insert *so much additional logic* to parse all those references that it would imho slow down the enforcer unnecessarily and given the nature of IoT devices, resources are a crucial factor that must be taken into account, not to mention that the overall complexity of the enforcer itself would be badly hurt.
If the API had a simpler schema and all of them would adhere to it, now that would be different, but what's done is done for good so let's think of something that actually helps us, not boggles us down.

Any thoughts on the new specs? I made them after having a quick look on how the properties are written in the example swagger file but I might've missed something (hopefully not, but just want you to confirm it)

**Comment by nopbyte** [comment]
Hi @thundron,

I cannot see the way to do they mapping as we had it in MappingKeys as described in the readme. We need to be able to create values that are passed to the API of the PDP which are composed from both, fixed strings, as well as parameters parsed from the call.

To have more clarity you should also include modifications to the relevant parts of the README, otherwise we are left with an ambiguous specification. Which is what we wanted to avoid when Csaba and me started documenting the behaviour of the enforcer.

**Comment by thundron** [comment]
@nopbyte Totally forgot about the readme, you're right!

**Comment by thundron** [comment]
@nopbyte While I update the readme accordingly, the mapping is done much more easily specifying the path of the parameter itself so the enforcer doesn't need to know the whole lot of specs for each single API since the path to be followed to take the parameter value already lies in the enforcer specs themselves. You can see the file changed in the PR for the examples but basically:
let's say we have the `entityId` in the payload property of the body parameter of the body object of a POST call (I believe one of the calls was almost exactly like this)
we just write a `path` property in the `entityId` object of the enforcer specs like
```
- path: body.body.payload.id
```
So we don't need to dig down in the deep sea of each API call, reference, sub-reference and sub-sub-reference (there were quite some specs structured like that) because we already have all the info for the enforcer to work its way to the specified data at the specified path.

**Comment by nopbyte** [comment]
I see why you want to use something like body.field.subfield, because it is an object... but we still need to compose arguments with fixed strings, and possible multiple arguments. Further, we need to be able to look them up in different parts of the HTTP request (body, url , url query parameters)

For example if we had:

 field: actions.{deviceId}.{command}
 
In the case that arguments come from the URL ( “/device/myId123/execute/on” ) this needs to resolve to the string "actions.myId123.on" based on the swagger specification and the mapping provided.

I don't see how you can achieve something like this. This requirement requires you to parse more than one part of the URL, and to be able to specify an arbitrary way to combine it with a static string, i.e. "actions.".

Is this possible in your scenario?

**Comment by thundron** [comment]
@nopbyte Yes it is, again there are some examples but here's the full explanation:
should we need to define a dynamic option made of fixed strings + variables, we can use the exact structure proposed above: I didn't in fact mention it, but I made it an array on purpose so that we can specify for example the `field` like this (there should be an example in the modified file too):
```yml
field: # list of strings | objects{path: string}
        - actions # simple string, won’t get parsed
        - protocol # simple string, won’t get parsed
        - path: body.payload.overview.protocol # object with path property where to look the value in
```
or simplified since the first two values are fixed:
```yml
field: # list of strings | objects{path: string}
        - actions.protocol # simple string, won’t get parsed
        - path: body.payload.overview.protocol # object with path property where to look the value in
```
so this will transalate into
```javascript
{
  // ... ,
  field: "actions.protocol.myAwesomeValue"
}
```
joining all array values with a `.`

**Comment by thundron** [comment]
@nopbyte In your example, I guess we'd have something along the lines of
```yml
field:
  - actions
  - path: url.deviceId
  - path: url.command
```

**Comment by thundron** [comment]
Again, with this being already in the enforcer spec, we totally skip the yml definitions which is a *huge* gain in both performance speed and resources optimization (since we don't have to access files nor keep them in memory after accessing them the first time)

**Comment by thundron** [comment]
Beware though: we will still need the `operationId` in order to easily map the API calls in the enforcer, the other route would be inserting the whole url `path` (e.g. `/devices/{id}/...`) which would be much more work for you than just writing the mapped operationId and passing it down in a header (best place in order to avoid messing up functionality when forwarding the request after it's been authorized)

**Comment by thundron** [comment]
The focus of this PR is basically this: the enforcer doesn't have to know the structure of the APIs (and conceptually, it *shouldn't*). Whoever integrates it on their module needs to know the APIs of his module and write them down in the enforcer specs, it should be fairly easy given that we document the README accordingly, but I wanted to first hear from you guys if this seems logical and maybe easier than having also you look down references and sub-references just to know what to write in the enforcer, as opposed to instead just knowing which properties you send down the body/url/headers and replicate them almost like you wrote them when writing the code itself

**Comment by thundron** [comment]
Readme's up to date, check it and see that everything's in place please.

**Comment by thundron** [comment]
The README is ready for review as well

**Comment by thundron** [comment]
@cskiraly @nopbyte ^

**Comment by nopbyte** [comment]
Hi,

so basically, you still use the swagger spec to inspect the API call provided by the operationId. The difference is that you added the url, body, etc keywords to avoid finding out where they are located but just go to the url (or the right part of the request) and look for the named element (e.g. deviceId)?

If I understood correctly, that sounds reasonable to me. I just have one more question, what do you mean by:

> which would be much more work for you than just writing the mapped operationId and passing it down in a header (best place in order to avoid messing up functionality when forwarding the request after it's been authorized)

It reads a bit as if you expect to modify the call received by the proxy to include some headers? but I am almost sure I misunderstood that part.

**Comment by thundron** [comment]
@nopbyte Exactly.
As for modifying the call sent to the enforcer: yes the best route I think is to include a new header `x-agile-api-method` in the API calls going to the enforcer (not the ones outgoing by the enforcer but the ones outgoing from services to the enforcer) that contains the `operationId` (since you're gonna need to modify the calls to refer to the `agile-enforcer` anyway).
Otherwise, how's that call recognized anyway? If the incoming request doesn't contain the operationId in *some* way, the enforcer can't map it to the correct one and having the full path in the enforcer configs instead of the operationId would force again the enforcer to parse the whole API specs and *assume* that what it finds is the correct API spec, which is prone to errors whilst being also resource hungry.

**Comment by thundron** [comment]
@nopbyte @cskiraly Any news on this?

**Comment by nopbyte** [comment]
@thundron this will not work. The whole point of this enforcer is to avoid modifying all the services.This is why in our initial discussion it was proposed to use the swagger spec. To be able to generically map calls.

If there are two API calls that have the same structure, for example a get to '/devices/{devideId}/status' without any parameters (for simplicity) and another one to '/devices/{devideId}/status' without any parameters too... Then the enforcer would not be able to choose which one to enforce. But this will NOT happen in our API. So, your second concern is not relevant, you can ignore this.

Even if this ever happens you can code the enforcer to pick one at random, but modifying every service is not viable. @cskiraly please comment on this if you disagree

**Comment by thundron** [comment]
@nopbyte First of all, the situation you described cannot happen **at all** because you can only have **one** API endpoint. It's impossible to have two calls with the exact same path *and* method, so we'll only ever have *one* `GET /devices/{deviceId}/status/`, *one* `POST /devices/{deviceId}/status` and so on. My concern was that without specifying the method in the configs too, I can't pick one or the other.
With that said, the swagger specs by themselves don't allow the enforcer (nor actually anything in existence without *some* mapping) to know which call I'm receiving. If you don't include the `operationId` in the call itself, it's just a random call with some parameters for the enforcer, and this is with or without the specs so I don't understand your disagreement: you'll *need* to modify the calls *anyway* if you want to ever include the enforcer in the modules.

**Comment by nopbyte** [comment]
In the readme we currently have this is written:

> if the enforcer gets a request to /agile-core/device/, it must be forwarded to /device/ to the agile-core sevice.

This will specify which swagger description should be parsed. You are including the component name (from docker in the url). That is the only change we can have. In this way, the enforcer maps the call to the specifications in the swagger file for that service. This is the key contribution from the enforcer: to be able to parse the specification, take the URL and see to which call it belongs.

**Comment by thundron** [comment]
@nopbyte So we're gonna have the `operationId` inside the URL right? Like `/agile-core/<operationId>/...`
I can't see how would you recognise an incoming call from `GET /agile-core/device/1/status` and map it to the right `operationId` if the operationId isn't found anywhere in the incoming call, so I'm definitely open to hear another solution if you have any

**Comment by cskiraly** [comment]
@thundron I'm open to simplifications of what we've proposed, but putting the `operationId` in the call as a header sounds counterintuitive to me. `operationId` is an internal reference used by swagger, which simply identifies an operation with a string, where operation is a <path,method> tuple.  I would prefer not modifying the HTTP calls with headers dependent on the internals of the services, when the path and method that is anyway present already identifies the operation you should look at. 
(for terminology, see https://swagger.io/docs/specification/paths-and-operations/)

**Comment by nopbyte** [comment]
I agree with @cskiraly 

Basically, what is needed is that the URL, query arguments, and the body are parsed in the enforcer.

Then, you need to remove the first part of the path, e.g. agile-core. 

with this string you can check which swagger spec applies for this call.

At this  point, I would start comparing the url to the paths in the swagger spec. One you find one that matches, you have to compare the arguments... Once you find one that matches you have your operation id.

Now, with the operation Id you go to the configuration of the enforcer and find out the mapping needed.

You construct the call and call the pdp.

If you get an OK, make the call. If not, return 403.

That is what is needed.

**Comment by cskiraly** [comment]
@thundron can you summarize what would be the goal of this change?

I'm trying to list some here that I can see as possibilities, with related comments:
1, if the goal is to change the notation used, I think that it can be OK. We just came up with one notation that seemed to have the descriptive power, but others could be fine as well (or even better).
2, if the goal is not to parse the `API spec` (swagger in yaml), only the configuration, that is I think doable. In that case the `configuration` should be more descriptive than in our example, as it should contain `<path,method>` to identify the matching entry.
3, anything else I miss?

**Comment by thundron** [comment]
@cskiraly Both and one more
1) Yes, the notation I proposed, whilst being a bit more code-alike which is not as visually appealing, is much less abstracted which means less work from the enforcer both on the development (= easier maintenance) and performance sides.
2) For the reason above, I was thinking of something similar otherwise we'd still need to parse each and every service's swagger file, which trust me is going to slow down performances a lot: consider how many calls the enforcer will need to authorize, even if it's 15ms slow down (being generous here), multiplied by 1000 calls it is 15000ms, which is a 15 *seconds* slow down also keeping the CPU fairly busy and we can't keep **all** the swagger files in-memory since they're quite big. If it was for just one service, I'd be down, but not with this architecture.
3) As I said, performances.

Instead of the `operationId` we can have the calls in sub-lists by method GET, POST, etc. so that
```yml
calls:
  get:
    /device/{devideId}/status:
      # see config changes proposed
```
So that I can at least:

1. check the method first, so we skip all other APIs with same path but different method
2. split the path
3. remove parameters at their index on both the source and the specs
4. check that every other segment is the same

cc @nopbyte

**Comment by nopbyte** [comment]
hi @thundron,

it looks to me that this may not be the only thing we need to check; for example, there may be some values in the body that determine whether you need to perform a different PDP call.

Anyway, this is taking too long to discuss, and getting ideas through is not working out. So, I talked to @cskiraly and we agree that you can go ahead and implement an approach that you are comfortable with, which can be showcased with a particular API from AGILE (ideally with agile-core).

That is what we would like to see, regardless of the format of the configuration. You can pick any format and document it...

Please let us know once you have something that can be demoed with the AGILE stack. This should do work.
